bhttp/src/http_compat/decode.rs

```diff
@@ -0,0 +1,538 @@
+    /// or an error if decoding fails.
+    pub async fn decode_message(self) -> Res<DecodeResult<R>> {
+        let mut async_message = Message::async_read(self.reader);
+
+        let header = async_message.header().await?;
+        let parts = HttpMessageParts::try_from(header)?;
+
+        Ok(DecodeResult {
+            async_message,
+            parts,
+        })
+    }
+}
+
+pub struct DecodeResult<R> {
+    async_message: AsyncMessage<R>,
+    parts: HttpMessageParts,
+}
+
+impl<R: FuturesAsyncRead + Unpin> DecodeResult<R> {
+    pub fn reader_ref(&self) -> &R {
+        self.async_message.reader_ref()
+    }
+}
+
+impl<R: FuturesAsyncRead + Unpin + Send + 'static> DecodeResult<R> {
+    pub fn into_full_message(self) -> Res<HttpMessage<R>> {
+        match self.parts {
+            HttpMessageParts::Request(parts) => Ok(HttpMessage::Request(Request::from_parts(
+                parts,
+                BhttpBody::new(self.async_message),
+            ))),
+            HttpMessageParts::Response(parts) => Ok(HttpMessage::Response(Response::from_parts(
+                parts,
+                BhttpBody::new(self.async_message),
+            ))),
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for HttpMessageParts {
+    type Error = Error;
+
+    /// Convert a BHTTP header into either HTTP request or response parts
+    ///
+    /// This method parses the control data and field section from a BHTTP header
+    /// and constructs the corresponding HTTP request or response parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - The BHTTP header to convert
+    ///
+    /// # Returns
+    ///
+    /// Either HTTP request parts or response parts, depending on the control data,
+    /// or an error if parsing fails.
+    fn try_from(header: crate::Header) -> Res<Self> {
+        if header.control().is_request() {
+            Ok(HttpMessageParts::Request(request::Parts::try_from(header)?))
+        } else {
+            Ok(HttpMessageParts::Response(response::Parts::try_from(
+                header,
+            )?))
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for request::Parts {
+    type Error = Error;
+
+    fn try_from(header: crate::Header) -> Res<Self> {
+        let control_data = header.control();
+
+        // Build the HTTP request parts
+        let mut builder = request::Builder::new();
+
+        // Set method
+        if let Some(method) = control_data.method() {
+            builder = builder.method(method);
+        }
+
+        // Convert field section to headers
+        let headers = HeaderMap::try_from(&header.fields)?;
+
+        // Set URI
+        let uri = {
+            let mut uri_builder = Uri::builder();
+
+            if let Some(scheme) = control_data.scheme() {
+                if !scheme.is_empty() {
+                    uri_builder = uri_builder.scheme(scheme);
+                }
+            }
```

Comment:
Again, the scheme is mandatory, so you can safely do something like this:


```suggestion
            let Some(scheme) = control_data.scheme() else {
                return Err(Error::Something);
            }
            if scheme.is_empty() {
                return Err(Error::Something);
            }
            uri_builder = uri_builder.scheme(scheme);
```